crates/ty_server/tests/e2e/pull_diagnostics.rs

```diff
@@ -639,3 +600,329 @@ fn sort_workspace_report_items(items: &mut [WorkspaceDocumentDiagnosticReport])
 
     items.sort_unstable_by(|a, b| item_uri(a).cmp(item_uri(b)));
 }
+
+#[test]
+fn workspace_diagnostic_long_polling_responds_on_shutdown() -> Result<()> {
+    let _filter = filter_result_id();
+
+    let workspace_root = SystemPath::new("src");
+    let file_path = SystemPath::new("src/test.py");
+    let file_content = "\
+def hello() -> str:
+    return \"world\"
+";
+
+    // Create a project with one file but no diagnostics
+    let mut server = create_workspace_server_with_file(workspace_root, file_path, file_content)?;
+
+    // Make a workspace diagnostic request to a project with one file but no diagnostics
+    // This should trigger long-polling since the project has no diagnostics
+    let request_id = send_workspace_diagnostic_request(&mut server);
```

Comment:
Can we update the `workspace_diagnostic_request` method on `TestServer` with more parameters like the progress params, partial result params, etc. instead of adding another function here?

Reply:
Oh, I now remember why it is necessary. The key difference is that this method doesn't await the response because we need to handle the response specially (it's long polling)